Support client create on Windows

Allow for using the client create on Windows (https://github.com/containerd/rust-extensions/tree/main/crates/client). Currently it fails to build with:

```
error[E0432]: unresolved import `tokio::net::UnixStream`
  --> C:\Users\jstur\.cargo\registry\src\index.crates.io-6f17d22bba15001f\containerd-client-0.4.0\src\lib.rs:81:9
   |
81 |     use tokio::net::UnixStream;
   |         ^^^^^^^^^^^^^^^^^^^^^^ no `UnixStream` in `net`
   |
note: found an item that was configured out
  --> C:\Users\jstur\.cargo\registry\src\index.crates.io-6f17d22bba15001f\tokio-1.32.0\src\net\mod.rs:51:27
   |
51 |     pub use unix::stream::UnixStream;
   |                           ^^^^^^^^^^

For more information about this error, try `rustc --explain E0432`.
error: could not compile `containerd-client` (lib) due to previous error
make: *** [Makefile:32: build-wasm] Error 101
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -46,13 +46,13 @@ jobs:
     steps:
       - uses: actions/checkout@v4
       - run: ./scripts/install-protobuf.sh
-      - run: cargo check --examples --tests -p containerd-shim -p containerd-shim-protos
+      - run: cargo check --examples --tests -p containerd-shim -p containerd-shim-protos -p containerd-client
       
       - run: rustup toolchain install nightly --component rustfmt
-      - run: cargo +nightly fmt -p containerd-shim -p containerd-shim-protos -- --check --files-with-diff
+      - run: cargo +nightly fmt -p containerd-shim -p containerd-shim-protos -p containerd-client -- --check --files-with-diff
       
       - run: cargo clippy -p containerd-shim -p containerd-shim-protos -- -D warnings
-      - run: cargo doc --no-deps -p containerd-shim -p containerd-shim-protos
+      - run: cargo doc --no-deps -p containerd-shim -p containerd-shim-protos -p containerd-client
         env:
           RUSTDOCFLAGS: -Dwarnings
 
@@ -80,7 +80,7 @@ jobs:
         env:
           # runc::tests::test_exec needs $XDG_RUNTIME_DIR to be set
           XDG_RUNTIME_DIR: /tmp/dummy-xdr
-      - run: cargo test -p containerd-shim -p containerd-shim-protos
+      - run: cargo test -p containerd-shim -p containerd-shim-protos -p containerd-client
         if: ${{ contains(matrix.os, 'windows') }}
 
   # Collect build timings
@@ -219,6 +219,12 @@ jobs:
           cargo run -p containerd-shim-protos --example shim-proto-connect \\.\pipe\containerd-shim-17630016127144989388-pipe
           $skeleton = get-process skeleton -ErrorAction SilentlyContinue
           if ($skeleton) { exit 1 }
+      - name: Run client
+        run: |
+          $ErrorActionPreference = "Stop"
+
+          get-service containerd
+          cargo run -p containerd-client --example version
 
   # Currently Github actions UI supports no masks to mark matrix jobs as required to pass status checks.
   # This means that every time version of Go, containerd, or OS is changed, a corresponding job should

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -20,6 +20,7 @@ jobs:
       - uses: actions/checkout@v4
 
       - run: ./scripts/install-protobuf.sh
+        shell: bash
 
       - run: cargo check --examples --tests --all-targets
       - run: cargo check --examples --tests --all-targets --all-features
@@ -46,6 +47,7 @@ jobs:
     steps:
       - uses: actions/checkout@v4
       - run: ./scripts/install-protobuf.sh
+        shell: bash
       - run: cargo check --examples --tests -p containerd-shim -p containerd-shim-protos -p containerd-client
       
       - run: rustup toolchain install nightly --component rustfmt
@@ -68,6 +70,7 @@ jobs:
     steps:
       - uses: actions/checkout@v4
       - run: ./scripts/install-protobuf.sh
+        shell: bash
       - run: |
           # runc-shim::cgroup::test_add_cgroup needs root permission to set cgroup
           mkdir -p /tmp/dummy-xdr
@@ -93,6 +96,7 @@ jobs:
     steps:
       - uses: actions/checkout@v4
       - run: ./scripts/install-protobuf.sh
+        shell: bash
       - run: cargo build --all-features --timings
       - uses: actions/upload-artifact@v4
         with:
@@ -181,6 +185,8 @@ jobs:
     steps:
       - name: Checkout extensions
         uses: actions/checkout@v4
+      - run: ./scripts/install-protobuf.sh
+        shell: bash
 
       - name: Install containerd
         run: |

--- a/crates/client/Cargo.toml
+++ b/crates/client/Cargo.toml
@@ -32,6 +32,9 @@ tower = { workspace = true, optional = true }
 [build-dependencies]
 tonic-build.workspace = true
 
+[dev-dependencies]
+tokio = { workspace = true, features = ["rt", "macros"]}
+
 [features]
 connect = ["tokio", "tower"]
 docs = []

--- a/crates/client/README.md
+++ b/crates/client/README.md
@@ -9,6 +9,8 @@ This crate implements a GRPC client to query containerd APIs.
 
 ## Example
 
+Run with `cargo run --example version`
+
 ```rust
 use containerd_client::{connect, services::v1::version_client::VersionClient};
 

--- a/crates/client/examples/version.rs
+++ b/crates/client/examples/version.rs
@@ -19,9 +19,13 @@ use containerd_client::Client;
 /// Make sure you run containerd before running this example.
 #[tokio::main(flavor = "current_thread")]
 async fn main() {
-    let client = Client::from_path("/var/run/containerd/containerd.sock")
-        .await
-        .expect("Connect failed");
+    #[cfg(unix)]
+    let path = "/var/run/containerd/containerd.sock";
+
+    #[cfg(windows)]
+    let path = r"\\.\pipe\containerd-containerd";
+
+    let client = Client::from_path(path).await.expect("Connect failed");
 
     let resp = client
         .version()

--- a/crates/client/src/lib.rs
+++ b/crates/client/src/lib.rs
@@ -82,7 +82,6 @@ pub mod events {
 pub async fn connect(
     path: impl AsRef<std::path::Path>,
 ) -> Result<tonic::transport::Channel, tonic::transport::Error> {
-    use tokio::net::UnixStream;
     use tonic::transport::Endpoint;
 
     let path = path.as_ref().to_path_buf();
@@ -91,7 +90,18 @@ pub async fn connect(
     let channel = Endpoint::try_from("https://[::]")
         .unwrap()
         .connect_with_connector(tower::service_fn(move |_| {
-            UnixStream::connect(path.clone())
+            #[cfg(unix)]
+            {
+                tokio::net::UnixStream::connect(path.clone())
+            }
+
+            #[cfg(windows)]
+            {
+                let client = tokio::net::windows::named_pipe::ClientOptions::new()
+                    .open(path.clone())
+                    .map_err(|e| std::io::Error::from(e));
+                async move { client }
+            }
         }))
         .await?;
 

--- a/rust-toolchain.toml
+++ b/rust-toolchain.toml
@@ -1,3 +1,3 @@
 [toolchain]
 channel = "1.71"
-components = ["rustfmt", "clippy", "llvm-tools-preview"]
+components = ["rustfmt", "clippy", "llvm-tools"]

--- a/scripts/install-protobuf.sh
+++ b/scripts/install-protobuf.sh
@@ -1,6 +1,7 @@
 #!/bin/bash
 
 # Helper script for Github Actions to install protobuf on different runners.
+echo "OS: $RUNNER_OS"
 
 if [ "$RUNNER_OS" == 'Linux' ]; then
     # Install on Linux

--- a/scripts/install-protobuf.sh
+++ b/scripts/install-protobuf.sh
@@ -11,7 +11,7 @@ elif [ "$RUNNER_OS" == 'macOS' ]; then
     brew install protobuf
 elif [ "$RUNNER_OS" == 'Windows' ]; then
     # Install on Windows
-    choco install protoc
+    choco install -y protoc
 else
     echo "Unsupported OS: $RUNNER_OS"
     exit 1

